Feat/Scrum 131 회원가입 도메인 & 프레젠테이션 로직 구현

derrickkim0109 opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경

<!-- 해당 PR에서 개발한 내용에 대해 알아야 하는 배경지식을 작성해주세요. -->

- 온보딩 flow에서 있는 각 뷰의 인터렉션이 정상 작동하도록 만들어야 합니다.

## 목표

- 약관 동의 화면에서 각 동의가 클릭 될 시 전체 동의도 클릭되고 다음 버튼 활성화
- 성별 및 출생년 선택 시 다음 버튼 활성화
- 프로필 텍스트 입력 시 다음 버튼 활성화
- Login API 정상호출 
- UserSignUpRequest 수정
- 다음 화면으로 이동 Coordinator Onboarding에서 관리할 수 있도록 구현
- Combine Cocoa로 뷰 바인딩 
- SignUp 로직 추가
- 자동로그인 기능 정상화

## 결과 (Optional)

<!-- 개발한 내용의 결과를 설명해주세요. !-->

- 각 인터렉션에 맞게 다음 버튼이 활성화 되고 다음 화면으로 넘어갈 수 있도록 coordinator에서 각 메서드를 통해 처리됩니다.

<!-- 
gif, 사진을 위한 Table

| | | |
|--|--|--|
|<img src="" width="250">|<img src="" width="250">|<img src="" width="250">|
-->

## 질문

- 토큰 만료라는 에러가 발생해서 자동로그인 API가 정상작동하지 않고 있어요. 이 부분은 추후 API가 정상화 되고 진행할까하는데 어떻게 생각하세요?

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
docs: issue templates 업데이트

derrickkim0109 committed:
remove: Feature Onboarding View 파일 폴더링

derrickkim0109 committed:
feat<SharedDesignSystem>: NavigationBar View 구현

derrickkim0109 committed:
feat<SharedUtil>: CocoaCombine 추가 구현

- UICollectionView didSelect, didDeselectItemAt
- UITextField

derrickkim0109 committed:
chore: 불필요한 개행 및 코드 제거

derrickkim0109 committed:
chore: 불필요한 개행 및 코드 제거

derrickkim0109 committed:
refactor: FeelinAPI 부분 수정

derrickkim0109 committed:
feat: Sign Up을 위한 Entity Model 구현

derrickkim0109 committed:
refactor: Login ViewController 로그인 Error 타입 변경

derrickkim0109 committed:
refactor: UserSignUpRequest init 추가 구현

derrickkim0109 committed:
refactor: UserSignUpRequest init 추가 구현

derrickkim0109 committed:
feat: UseAgreementVC binding 구현 및 Navigation Bar 추가

derrickkim0109 committed:
feat: UserInformationVC binding 구현 및 Navigation Bar 추가

derrickkim0109 committed:
refactor: ProfileVC binding 구현 및 navigation Bar 추가

derrickkim0109 committed:
feat: Onboarding Coordinator 추가 구현

derrickkim0109 commented on Projects/Coordinator/Main/Interface/Sources/MainInterface.swift:
```diff
@@ -19,7 +19,6 @@ public final class MainCoordinator: Coordinator {
 
 private extension MainCoordinator {
     func configureMainController() {
-        let viewController = MainRootViewController()
-        navigationController.pushViewController(viewController, animated: false)
```
사용하지 않는 코드라 제거했습니다.

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/UserSignUpRequest.swift:
```diff
@@ -27,6 +27,12 @@ public struct Term: Encodable {
     let title: String
     let agreement: String
+
+    public init(agree: Bool, title: String, agreement: String) {
+        self.agree = agree
+        self.title = title
+        self.agreement = agreement
+    }
```
사용처에서 initializer를 사용해야 해서 구현했습니다.

derrickkim0109 committed:
refactor: UserSignUpRequest 프로퍼티 네이밍 변경

derrickkim0109 committed:
refactor: UserSignUpRequest 프로퍼티 네이밍 변경

derrickkim0109 commented on Projects/Domain/OAuth/Interface/Sources/Model/UserSignUpEntity.swift:
```diff
@@ -0,0 +1,53 @@
+import Core
+
+public struct UserSignUpEntity {
+    public var socialAccessToken: String?
+    public var oAuthType: OAuthType?
+    public var nickname: String?
+    public var gender: GenderEntity?
+    public var birthYear: String?
+    public var terms: [Term]
+
+    public init(
+        socialAccessToken: String? = nil,
+        oAuthType: OAuthType? = nil,
+        nickname: String? = nil,
+        gender: GenderEntity? = nil,
+        birthYear: String? = nil,
+        terms: [Term] = []
+    ) {
+        self.socialAccessToken = socialAccessToken
+        self.oAuthType = oAuthType
+        self.nickname = nickname
+        self.gender = gender
+        self.birthYear = birthYear
+        self.terms = terms
+    }
```
논의한 대로 Onboarding 화면 별로 나눠서 데이터를 이동시켜야 해서 Optional로 수정했습니다~!

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/Profile/ProfileView.swift:
```diff
@@ -14,21 +14,34 @@ import PinLayout
+
     // MARK: - components
-    
+    private let navigationBar = NavigationBar()
+
+    lazy var backButton: UIButton = {
+        let button = UIButton()
+        button.translatesAutoresizingMaskIntoConstraints = false
+
+        let userInterfaceStyle = traitCollection.userInterfaceStyle
+        let image = userInterfaceStyle == .light ? FeelinImages.backLight : FeelinImages.backDark
+        button.setImage(image, for: .normal)
+
+        return button
+    }()
+
```
NavigationBar 뷰에 백버튼을 주입하는 방식으로 구현했습니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/Profile/ProfileView.swift:
```diff
@@ -37,68 +50,78 @@ final class ProfileView: UIView {
-    private lazy var nicknameTextField = FeelinLineInputField(placeholder: "닉네임")
-    
+
+    lazy var nicknameTextField = FeelinLineInputField(placeholder: "닉네임")
```
해당 Custom TextField의 TextField에 접근하기 위해 internal로 변경했습니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/UserAgreement/CheckBoxButton.swift:
```diff
@@ -14,13 +14,13 @@ final class CheckBoxButton: UIButton {
-    private (set) public var isChecked: Bool = false {
+
+    public override var isSelected: Bool {
         didSet {
             updateImage()
         }
     }
-    
+
```
isChecked를 기존에 있는 프로퍼티인 isSelected로 대체했습니다.

inwoodev replied:
맞습니다~ isChecked는 불필요한 값이었어요 ㅎㅎ

derrickkim0109 committed:
refactor: 불필요한 추가 설정 제거

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/UserAgreement/UserAgreementListView.swift:
```diff
@@ -7,163 +7,154 @@
-            default:
-                return
-            }
+            button.configuration?.baseForegroundColor = button.state == .selected ? nil : Colors.gray04
+            button.configuration?.baseBackgroundColor = button.state == .selected ? Colors.alertSuccess : Colors.background
+            self?.allAgreeLabel.textColor = button.state == .selected ? Colors.alertSuccess : Colors.gray04
```
기존 코드를 삼항연산자로 대체했습니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/UserAgreement/UserAgreementListView.swift:
```diff
@@ -7,163 +7,154 @@
     }()
-    
+
     lazy var serviceUsageAgreeButton: UIButton = {
-        return self.createAgreementButton("서비스 이용약관 동의")
+        return self.createAgreementButton(TermEntity.serviceUsage.title)
     }()
-    
+
     lazy var personalInfoAgreeButton: UIButton = {
-        return self.createAgreementButton("개인정보처리방침 동의")
+        return self.createAgreementButton(TermEntity.personalInfo.title)
     }()
-    
+
```
title, agreement에 대한 부분은 따로 관리가 필요하다고 생각해서 TermEntity를 구현해 각 title의 내용을 관리할 수 있도록 구현했습니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/UserInformationView.swift:
```diff
@@ -17,6 +17,28 @@ final class UserInformationView: UIView {
     
+    private let navigationBar = NavigationBar()
+    
+    lazy var backButton: UIButton = {
+        let button = UIButton()
+        button.translatesAutoresizingMaskIntoConstraints = false
+
+        let userInterfaceStyle = traitCollection.userInterfaceStyle
+        let image = userInterfaceStyle == .light ? FeelinImages.backLight : FeelinImages.backDark
+        button.setImage(image, for: .normal)
+
+        return button
+    }()
+
+    let skipButton: UIButton = {
+        let button = UIButton()
+        button.translatesAutoresizingMaskIntoConstraints = false
+        button.setTitle("건너뛰기", for: .normal)
+        button.titleLabel?.font = SharedDesignSystemFontFamily.Pretendard.regular.font(size: 16)
+        button.setTitleColor(Colors.gray05, for: .normal)
+        return button
+    }()
```
navigationBar에 backButton과 skipButton을 넣습니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/Components/UserInformationView.swift:
```diff
@@ -79,9 +101,17 @@ final class UserInformationView: UIView {
+
             flex.addItem(titleLabel)
-                .marginTop(72)
+                .marginTop(28)
```
네비게이션 바의 추가 구현으로 해당 값을 수정했ㅅ브니다.

derrickkim0109 commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/EditProfileViewController.swift:
```diff
@@ -6,87 +6,72 @@
+            }
+            .store(in: &cancellables)
+
+        profileCharacterCollectionView.publisher(for: [.didSelectItem, .didDeselectItem])
+            .sink { [weak self] indexPath in
+                guard let self = self else { return }
+
+                if let cell = self.profileCharacterCollectionView.cellForItem(at: indexPath) as? ProfileCharacterCell {
+                    let isSelected = self.profileCharacterCollectionView.indexPathsForSelectedItems?.contains(indexPath) ?? false
+                    cell.setSelected(isSelected)
+                    selectButton.isEnabled = isSelected
+                    selectedProfileIndex = indexPath.row
+                }
+            }
+            .store(in: &cancellables)
```
collectionView도 combine cocoa가 될 수 있도록 변경했습니다.

derrickkim0109 commented on Projects/Shared/DesignSystem/Sources/Views/FeelinLineInputField.swift:
```diff
@@ -31,7 +31,13 @@ public final class FeelinLineInputField: UIView {
     }()
-    
+
+    public var isValid: Bool = true {
+        didSet {
+            separator.backgroundColor = isValid ? Colors.gray02 : Colors.alertWarning
+        }
+    }
```
잘못 작성된 텍스트라면 다음 버튼이 활성화되지 않도록 하기 위해 해당 프로퍼티를 추가했습니다.

derrickkim0109 reviewed (commented):
추가적으로 수정이 필요한 부분 말씀해주시면 감사하겠습니다~!

inwoodev commented on Projects/Shared/DesignSystem/Sources/Views/NavigationBar.swift:
```diff
@@ -0,0 +1,68 @@
+public extension NavigationBar {
+    func addTitleView(_ contentView: UIView) {
+        titleView.flex.define { flex in
+            flex.addItem(contentView).grow(1).shrink(1).marginHorizontal(5)
+        }
+        titleView.flex.layout()
+    }
+
+    func addLeftBarView(_ contentView: UIView) {
+        leftBarView.flex.define { flex in
+            flex.addItem(contentView).width(40).height(40)
+        }
+        leftBarView.flex.layout()
+    }
+
+    func addRightBarView(_ contentView: UIView) {
+        rightBarView.flex.define { flex in
+            flex.addItem(contentView).width(60).height(40)
+        }
+        rightBarView.flex.layout()
+    }
+}
+
+private extension NavigationBar {
+    func setupDefault() {
+        rootFlexContainer.flex.direction(.row).alignItems(.center)
+        leftBarView.flex.width(40).height(40)
+        titleView.flex.grow(1).alignItems(.center).justifyContent(.center)
+        rightBarView.flex.width(60).height(40)
+    }
+
+    func addUIComponents() {
+        addSubview(rootFlexContainer)
+        rootFlexContainer.flex.define { flex in
+            flex.addItem(leftBarView)
+            flex.addItem(titleView).grow(1).shrink(1)
+            flex.addItem(rightBarView)
+        }
+    }
+}
```
고생하셨습니다~ 요긴하게 사용하겠네요~!!

inwoodev commented on Projects/Coordinator/Onboarding/Interface/Sources/OnboardingInterface.swift:
```diff
@@ -61,7 +61,36 @@ private extension OnboardingCoordinator {
+        viewController.coordinator = self
+        navigationController.pushViewController(viewController, animated: true)
+    }
+
```
네비게이션 같은 경우 show동사를 활용하는걸까요?
그렇다면 modal 같은 경우 어떤 동사를 활용해 볼 수 있을까요~?

derrickkim0109 replied:
이 부분은 push~로 변경되어있습니다.

inwoodev commented on Projects/Shared/Util/Sources/Extensions/UICollectionView+.swift:
```diff
@@ -22,3 +23,83 @@ extension UICollectionView {
+    private let events: [UICollectionView.PublisherEvent]
+
+    init(
+        subscriber: S,
+        collectionView: UICollectionView,
+        events: [UICollectionView.PublisherEvent]
+    ) {
+        self.subscriber = subscriber
+        self.collectionView = collectionView
+        self.events = events
+        super.init()
+        self.collectionView?.delegate = self
+    }
+
+    /// Subscribers.Demand는 Combine의 백프레셔(Backpressure)를 제어하기 위한 것
+    /// 여기서는 기본적으로 무시하고 이벤트가 발생할 때마다 전송하는 방식으로 구현
+    func request(_ demand: Subscribers.Demand) { }
+
+    func cancel() {
+        subscriber = nil
+    }
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        didSelectItemAt indexPath: IndexPath
+    ) {
+        if events.contains(.didSelectItem) {
+            _ = subscriber?.receive(indexPath)
+        }
+    }
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        didDeselectItemAt indexPath: IndexPath
+    ) {
+        if events.contains(.didDeselectItem) {
+            _ = subscriber?.receive(indexPath)
+        }
+    }
+}
```
lgtm~~👍👍👍

derrickkim0109 committed:
refactor: 화면 전환 메서드 show -> push로 네이밍 변경

derrickkim0109 committed:
feat: FeelinAPI signUp case 추가

derrickkim0109 committed:
feat: FeelinAPI signUp case 추가

derrickkim0109 committed:
feat: FeelinAPI signUp case 추가

derrickkim0109 committed:
test: NetworkProvider SignUp 테스트 코드 구현

derrickkim0109 committed:
test: NetworkProvider SignUp 테스트 코드 구현

derrickkim0109 committed:
test: NetworkProvider SignUp 테스트 코드 구현

derrickkim0109 committed:
test: NetworkProvider SignUp 테스트 코드 구현

derrickkim0109 committed:
refactor: 닉네임 텍스트 정규식 수정

derrickkim0109 committed:
feat: Welcome View 구현

- coordinator 연동 (WelcomeViewController push 및 TabBarCoordinator로 이동)

derrickkim0109 committed:
refactor: Login Logic 수정

- error로 없는 user일 경우 회원가입으로 가는 Flow
- MockRecentLoginRecordService 추가
- accessToken, refreshToken 이름 수정

derrickkim0109 committed:
refactor: Login Logic 수정

- error로 없는 user일 경우 회원가입으로 가는 Flow
- MockRecentLoginRecordService 추가
- accessToken, refreshToken 이름 수정

derrickkim0109 committed:
refactor: Login Logic 수정

- error로 없는 user일 경우 회원가입으로 가는 Flow
- MockRecentLoginRecordService 추가
- accessToken, refreshToken 이름 수정

derrickkim0109 committed:
refactor: Login Logic 수정

- error로 없는 user일 경우 회원가입으로 가는 Flow
- MockRecentLoginRecordService 추가
- accessToken, refreshToken 이름 수정

derrickkim0109 committed:
feat: UserSignUp UseCase 추가

- SignUpService
- SignUpError
- FeelinError
- Request, Entity 객체 수정

derrickkim0109 committed:
feat: UserSignUp UseCase 추가

- SignUpService
- SignUpError
- FeelinError
- Request, Entity 객체 수정

derrickkim0109 committed:
feat: ProfileViewModel 추가

derrickkim0109 committed:
feat: ProfileViewModel 추가

derrickkim0109 committed:
feat: ProfileViewModel 추가

derrickkim0109 committed:
refactor: ProfileViewController에 ViewModel 주입

derrickkim0109 committed:
refactor: ProfileViewController에 ViewModel 주입

derrickkim0109 committed:
refactor: ProfileViewController에 ViewModel 주입

derrickkim0109 committed:
refactor: ProfileViewController에 ViewModel 주입

derrickkim0109 committed:
refactor: 년생 데이터 "년" 문구 분리

derrickkim0109 committed:
feat: ProfileViewController SignUp 로직 및 UI Interaction 추가

derrickkim0109 committed:
feat: ProfileViewController SignUp 로직 및 UI Interaction 추가

derrickkim0109 committed:
feat: ProfileViewController SignUp 로직 및 UI Interaction 추가

derrickkim0109 committed:
refactor: 자동로그인 로직 수정

derrickkim0109 committed:
chore: 불필요한 부분 제거

derrickkim0109 committed:
refactor: Network feelinAPIError로 수정되어 userNotFound시 case로 변경

derrickkim0109 committed:
refactor: Network feelinAPIError로 수정되어 userNotFound시 case로 변경

derrickkim0109 committed:
refactor: SignUp API 실패시 Alert 추가

derrickkim0109 committed:
refactor: Request 객체 프로퍼티 타입 변경

inwoodev commented on Projects/Core/Network/Tests/Sources/NetworkProviderTests.swift:
```diff
@@ -220,4 +220,70 @@ final class NetworkProviderTests: XCTestCase {
+                headerFields: nil
+            )!
+
+            let mockData = """
+            {
+                "status": "회원가입이 완료되었습니다.",
+                "data": {
+                    "accessToken": "\(expectedAccessToken)",
+                    "refreshToken": "\(expectedRefreshToken)"
+                }
+            }
+            """.data(using: .utf8)!
```
태현님, 요거 모델 수정해야할듯 해요!! 저희 status는 안받기로 합의된걸로 기억합니당~

derrickkim0109 replied:
수정했습니다.

inwoodev commented on Projects/DependencyInjection/Sources/DIContainer/DIContainer.swift:
```diff
@@ -88,3 +88,18 @@ public extension DIContainer {
+
+// MARK: SignUp
+
+public extension DIContainer {
+    static func registerSignUpService() {
+        standard.register(.signUpService) { resolver in
+            let tokenStorage = try resolver.resolve(.tokenStorage)
+            let networkProvider = try resolver.resolve(.networkProvider)
+            return SignUpService(
+                networkProvider: networkProvider,
+                tokenStorage: tokenStorage
+            )
+        }
+    }
+}
```
요렇게 가시죠~ 저도 수정하겠습니다~

inwoodev commented on Projects/Domain/OAuth/Sources/Service/UserValidityService.swift:
```diff
@@ -6,17 +6,26 @@
 //
 
 import Combine
+import Core
 import Foundation
 import CoreNetworkInterface
 import DomainOAuthInterface
```
Core가 있다면 CoreNetworkInterface는 굳이 import 안 해도 될듯 해요~

derrickkim0109 replied:
수정했습니다~

derrickkim0109 committed:
refactor: 자동로그인 리턴 타입 수정 및 accessToken으로 변경

derrickkim0109 committed:
refactor: 로그인 실패시 알림 문구 변경

derrickkim0109 committed:
refactor: 로그인 실패시 알림 문구 변경

derrickkim0109 committed:
refactor: ProfileViewController Input, output 구조 변경

derrickkim0109 committed:
refactor: ProfileViewController Input, output 구조 변경

derrickkim0109 committed:
refactor: Feelin Default Response 추가

derrickkim0109 committed:
refactor: Error Code 추가

derrickkim0109 committed:
refactor: Error Code 추가

derrickkim0109 commented:
- Auto Login에 필요한 Response 객체 추가
- ProfileViewController Input, Output 구조 변경
- Sign up 비지니스 로직 추가

derrickkim0109 committed:
chore: 불필요한 부분 제거

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -73,52 +58,43 @@ public final class ProfileViewController: UIViewController {
-                editProfileViewController.modalPresentationStyle = .overFullScreen
-                present(editProfileViewController, animated: false)
+                self.editProfileViewController.modalPresentationStyle = .overFullScreen
+                self.present(self.editProfileViewController, animated: false)
```
`guard let self = self else { return }` 구문이 굳이 필요하지 않은 듯 해요~ p4

derrickkim0109 replied:
수정했습니다~

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -73,52 +58,43 @@ public final class ProfileViewController: UIViewController {
-                return false
-            }
-            .eraseToAnyPublisher()
+        let nicknameTextPublisher = nicknameTextField.textField.textPublisher
+            .share()
```
`.share()`가 어떤 측면에서 필요할까요~? p5

derrickkim0109 replied:
불필요해서 제거했습니다~!

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -73,52 +58,43 @@ public final class ProfileViewController: UIViewController {
+            nicknameTextPublisher: nicknameTextPublisher.eraseToAnyPublisher(),
+            profileImagePublisher: profileImagePublisher.eraseToAnyPublisher(),
+            nextButtonTapPublisher: nextButton.publisher(for: .touchUpInside).eraseToAnyPublisher()
         )
```
디코로 말씀드린거 같이 하나의 publisher를 input에 주입하는 방향으로 수정하셔도 좋을 듯 해요~ p3

derrickkim0109 replied:
수정했습니다~!

derrickkim0109 committed:
refactor: 불필요한 Publisher 제거 및 수정

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewModel/ProfileViewModel.swift:
```diff
@@ -31,33 +38,65 @@ public final class ProfileViewModel {
+                return image
+            }
+            .eraseToAnyPublisher()
+    }
```
요거 image는 optional이어도 될듯 한데 어떻게 생각하시나요~ p3

derrickkim0109 replied:
좋습니다~! 수정했습니다.

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewModel/ProfileViewModel.swift:
```diff
@@ -31,33 +38,65 @@ public final class ProfileViewModel {
+                input.nicknameTextPublisher,
+                input.profileImagePublisher
+            )
+            .compactMap { (nickname, profileCharacter) -> UserSignUpEntity? in
+                guard let nickname = nickname, !nickname.isEmpty, nickname.count < 10 else { return nil }
+                return UserSignUpEntity(nickname: nickname, profileCharacter: profileCharacter)
+            }
+            .eraseToAnyPublisher()
```
compactMap을 통해 optional한 값이 방출되는게 조금 마음에 걸립니다...ㅎ

derrickkim0109 replied:
수정했습니다.

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewModel/ProfileViewModel.swift:
```diff
@@ -31,33 +38,65 @@ public final class ProfileViewModel {
+            .flatMap { (tap, entity) in
+                return Just(SignUpResult.success).eraseToAnyPublisher()
             }
             .eraseToAnyPublisher()
```
CombineLatest 중첩인데 작동 잘 되는지 확인 요청 드려요~ p2

derrickkim0109 replied:
확인 결과 잘 작동합니다. zip으로 변경 시 nickname을 최신 값으로 가져올 수 없어 combineLatest 중첩이 필요합니다. 

혹시 다른 방법이 있을까요..?

inwoodev replied:
아닙니다~~ 확인차 코멘트 드린거에요. 잘 작동한다면 굳이 다른 방법을 찾지 않아도 될듯 합니다~

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewModel/ProfileViewModel.swift:
```diff
@@ -7,19 +7,26 @@
 
     private let signUpUseCase: SignUpUseCase
     private var userSignUpEntity: UserSignUpEntity
+    private var profileImageErrorPublisher = PassthroughSubject<String, Never>()
```
profileImageErrorPublisher가 꼭 필요한가 싶기는 합니다. api콜을 통해서 이미지를 가져오지 않기도 하고 이미지 갯수가 많은 것도 아니어서요~! p4

derrickkim0109 replied:
제거했습니다.

inwoodev reviewed (changes requested):
고생하셨습니다. 몇 가지 코멘트 확인 요청드려요~

derrickkim0109 committed:
refactor: FeelinDefalutResponse 프로퍼티 네임 변경

derrickkim0109 committed:
refactor: Profile Image Optional하게 받을 수 있도록 변경 및 불필요한 코드 제거

derrickkim0109 committed:
refactor: NetworkProviderTests Mock Data 변경

derrickkim0109 committed:
refactor: 회원가입 데이터 로직 수정

derrickkim0109 committed:
chore: 파라미터 빈값 처리

derrickkim0109 committed:
chore: 불필요한 코드 제거

derrickkim0109 committed:
refactor: Profile ViewModel Sign up 코드 수정

inwoodev commented on Projects/Coordinator/Onboarding/Interface/Sources/OnboardingInterface.swift:
```diff
@@ -55,10 +63,71 @@ private extension OnboardingCoordinator {
+
+    public func pushServiceUsageViewController() {
+        
+    }
+
+    public func pushPersonalInfoUsageViewController() {
+
+    }
+
+    public func pushUserInformationViewController(model: UserSignUpEntity) {
+        let viewController = UserInformationViewController(model: model)
+        viewController.coordinator = self
+        navigationController.pushViewController(viewController, animated: true)
+    }
+
+    public func pushProfileViewController(model: UserSignUpEntity) {
+        registerSignUpDependencies()
+        let viewModel = profileDependencies(model: model)
+        let viewController = ProfileViewController(viewModel: viewModel)
+        viewController.coordinator = self
+        navigationController.pushViewController(viewController, animated: true)
+    }
+
+    public func pushWelcomeViewController() {
+        let viewController = WelcomeViewController()
+        viewController.coordinator = self
+        navigationController.pushViewController(viewController, animated: true)
+    }
+
+    public func connectTabBarFlow() {
+        navigationController.popToRootViewController(animated: false)
+        let onboardingCoordinator = TabBarCoordinator(
+            navigationController: navigationController
+        )
+
+        onboardingCoordinator.delegate = self
+        onboardingCoordinator.start()
+        childCoordinators.append(onboardingCoordinator)
+    }
+    
```
👍👍

inwoodev reviewed (approved):
확인했습니다~ input & output로직도 공들이신만큼 개선되어보입니다. 머지하셔도 좋습니다~ 얼른 다음 feature로 넘어가시죠
